API updates

symroe opened the pull request:
I expect the tests to fail on this at the moment, as I've not updated them to track changes while I'm still making so many of them.

This is a first draft at a re-designed `next` API.

I have a WCIVF importer mostly working based on this version, so the initial sanity check has been passed, however it might be worth thinking about other changes we'd like to see now that we're changing so much.

Headline changes are:

* Removed a load of endpoints, including `posts`, `candidates_for_postcode`, `memberships`, `result_sets`, `candidate_results`, `party_sets`

* Moved to a "keyed by ballot paper ID" approach in other places, mainly moving `post_elections` to `ballots`.

*The `ballot` endpoint tried to be everything that is actually on a ballot as a base (nomination details, info about the ballot, including the election and "post").

* A ballot also includes SOPN documents and results, if we have them

* The `person` endpoint no longer includes versions – this is moved to a `people/[pk]/versions` sub-route. Same for `history` (`LoggedActions`) – This is a new feature, but one we talked about adding to the frontend soon anyway.

* People now include all `PersonIdentifiers`, in a list rather than a hash / dict. This enables us to move to `n` IDs per (person, type) in future, but might make use tricky before then? I always prefer a dict for getting objects, so maybe we have a dict for `value_type` and a list of `values`, ordered in some way that means e.g. person['twitter_username'][0] is the "most important". Feels like a wider topic than this PR, but still worth thinking about

* Ballots have some filters, but not enough at the moment. I wonder what we should add next?

* People have the recently updated filter, but no others. I expect something would be useful here, but not sure of use cases. Something that's easy enough to add on demand. 

* People now issues a 301 to another person if there's a `PersonRedirect` in place. Is this RESTful? I feel like it's reasonable, and more useful than throwing a 404 like it does at the moment

* I've moved API views to `[app_name]/api_views.py` files, and serializers to `[app_name]/serializers.py`. This is a trend started in https://github.com/DemocracyClub/yournextrepresentative/pull/633 – I think it's reasonable, but views welcome. This approach at the moment doesn't support versioning, so I wonder if I need to make them all packages with a sub-package per version? Not sure how to deal with this longer term, but I do like apps having their API gubbins in them somewhere.

* There is still a load of cleaning up around the cached API output, tests, live site importer and optimizing queries, but I don't want to get in to any of that until we have a broad agreement on the actual API design first.

chris48s commented:
## Chapter 1

There is more feedback for me to type up this afternoon, but this first point is large and important so I'll post this one in isolation. This just fleshes out and formalises what we already talked about on the call earlier:

As a result of removing the membership object, we're now mixing our metaphors a bit - there are certain properties which are sometimes properties of a candidate and sometimes properties of a ballot which is confusing. At the moment, we've got something like this:

```js
// /people/1.json

{
  "id": 1,
  "url": "/next/people/1.json",
  "name": "Aktar Beg",
  //bla
  "ballots": [
    {
      "ballot": "/next/ballots/parl.romsey-and-southampton-north.2010-05-06/",
      "elected": null,
      "party_list_position": null
      // side note: wierd that we have party_list_position, but not party
    }
  ]
}
```

and

```js
// /ballots/parl.romsey-and-southampton-north.2010-05-06.json

{
  "url": "/next/ballots/parl.romsey-and-southampton-north.2010-05-06.json",
  "ballot_paper_id": "parl.romsey-and-southampton-north.2010-05-06",
  //bla
  "candidates": [
    {
      "id": 61416, // side note: I think this is actually the membership id anyway, not the person id??
      "elected": null,
      "party_list_position": null,
      "name": "Aktar Beg",
      "party": {
        "ec_id": "PP53",
        "name": "Labour Party",
        "legacy_slug": "party:53"  // side note: is this really important in the short serializer?
      }
    }
  ]
}
```

To maintain more internal consistency, we should present both ballot and person as having a list of "candidacies" (the relationship between a person and a ballot). Fields which are a property of the candidacy should always be presented as top-level keys on the candidacy object. Properties relating to the person or ballot should be always serialized into a person or ballot object under the candidacy, so that would move us to something more like:

```js
// /people/1.json

{
  "id": 1,
  "url": "/next/people/1.json",
  "name": "Aktar Beg",
  //bla
  "candidacies": [
    {
      "ballot": {
        "ballot_paper_id": "parl.romsey-and-southampton-north.2010-05-06",
        "url": "/next/ballots/parl.romsey-and-southampton-north.2010-05-06/"
      }
      "party_list_position": null
      "party": {
        "url": "/next/ballots/parties/PP53.json",
        "ec_id": "PP53",
        "name": "Labour Party"
      }
    }
  ]
}
```

and

```js
// /ballots/parl.romsey-and-southampton-north.2010-05-06.json

{
  "url": "/next/ballots/parl.romsey-and-southampton-north.2010-05-06.json",
  "ballot_paper_id": "parl.romsey-and-southampton-north.2010-05-06",
  //bla
  "candidacies": [
    {
      "person": {
        "id": 1,
        "url": "/next/ballots/people/1.json",
        "name": "Aktar Beg",
      }
      "party_list_position": null,
      "party": {
        "url": "/next/ballots/parties/PP53.json",
        "ec_id": "PP53",
        "name": "Labour Party"
      }
    }
  ]
}
```

(accepting that you might want to add a field or 2 to ballot or person there).

That moves us to a position where:
- We use the terminology "candidacies" consistently across both endpoints
- things that are a property of a ballot/person/candidacy are always a property of a ballot/person/candidacy
- short-serialized entities always link to the detailed entity

One problem with what I've proposed above is that now a "candidacy" object requires 2 different serializers depending on where we're presenting it. This is going to inevitably lead to "drift" as we change things in one place and forget to change them in another. What we could do is define a single serializer for a candidacy, which always presents a candidacy as 


```js
{
  "person": {
    "id": 1,
    "url": "/next/ballots/people/1.json",
    "name": "Aktar Beg",
  }
  "ballot": {
    "ballot_paper_id": "parl.romsey-and-southampton-north.2010-05-06",
    "url": "/next/ballots/parl.romsey-and-southampton-north.2010-05-06/"
  }
  "party_list_position": null,
  "party": {
    "url": "/next/ballots/parties/PP53.json",
    "ec_id": "PP53",
    "name": "Labour Party"
  }
}
```

This is slightly less nice as we double up the person info if we're serialized in `/person` and we double up the ballot info if we're serialized in `/ballot`, but the tradeoff is having a single serializer for a candidacy means we change things in one place and both endpoints will be consistent, and that's a useful property to have too IMO.

The other thing I've not dealt with so far is results...

If we go down the route of maintaining a single serializer for a candidacy (regardless of where the object lives), I reckon the sensible decision here is that a candidacy object always has a full result object serialized in it. If we're going to go down the route of saying that a candidacy object has 2 different serializers depending on whether its a property of a person or a ballot object, it makes more sense to say they also have a different result serializer dependent on context: The one that gets used if the candidacy is a property of a person only includes `elected: true/false`, whereas if the candidacy is a property of a ballot then its a full result object.

Just thinking about results and the concept of getting rid of the results endpoint.. are there other things that make sense to think of as a property of a `result_set` (e.g: turnout, number of spoilt ballots, etc), or do you see those as all being things that should also be loaded into the ballot properties instead?

chris48s commented:
## Chapter 2

(I've not even looked at the code yet - lol)

> The `person` endpoint no longer includes versions – this is moved to a `people/[pk]/versions` sub-route. Same for `history` (`LoggedActions`) – This is a new feature, but one we talked about adding to the frontend soon anyway.

I have several questions about this.. I'm not sure I know what the answers are, but I feel like they're good things to think about:

Not all `LoggedAction` events are related to a person. For example:
* `suggest-ballot-lock`
* `constituency-lock` (badly named)
* `sopn-upload`

all relate to a ballot. Should `/ballots` also also have a `/history`, or is `/people` special? Are there other entities that a `LoggedAction` can relate to? Should they all have one? Would this be better thought of as a filter by entity on the `/logged_actions` endpoint?

How do I discover `/versions` , `/history` ? If they're going to exist, lets link them in the person object.

I _think_ my opinion is that `people/{id}/versions` makes sense because versions inherently relate to a person and its a massive object that we can usefully drop from the default serialized person object but `/history` might be better delivered as filters on `/logged_action`. Open to your thoughts on it though..

---

> Ballots have some filters, but not enough at the moment. I wonder what we should add next?

My nominations would be:

* date: "give me all ballots where polling day is `2019-05-02`". On Every Election, I did it so you can filter by a date range too: https://github.com/DemocracyClub/EveryElection/commit/e56fb08b6c65618eeec140628974617170bc5b77
* election: "give me all the ballots which are a child of `local.allerdale.2019-05-02`"
* current/future=1 (if we don't already have them)

---

> People now issues a 301 to another person if there's a `PersonRedirect` in place. Is this RESTful?

:+1:  It is in my opinion. REST is all about making use of HTTP primitives :man_shrugging: Sure you can find someone who will tell you it isn't though :trollface: 

---

`/party_registers` endpoint:

I have a weirdly large amount of feedback about an endpoint that returns such little output :laughing: 

* If I call http://127.0.0.1:8000/api/next/party_registers it works, but http://127.0.0.1:8000/api/next/party_registers.json throws an error. All the other endpoints work witth a `.json` suffix.

* On my local copy, the `party_registers` endpoint returns:

  ```json
  [
      "GB",
      "NI",
      null
  ]
  ```
  Is `null` a sensible value here. Have I done something silly to my local copy?


* Can we make this return a top-level object which is formatted the same way as the other endpoints i.e:

  ```json
  {
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
      {
        "slug": "GB"
      },
      {
        "slug": "NI"
      }
    ]
  }
  ```

  This buys us 2 things:

    * All the endpoints work the same way
    * If we ever need to add some other bit of info to it, we've got some flexibility. i.e: we can do:

  ```json
  {
    "count": 2,
    "next": null,
    "previous": null,
    "results": [
      {
        "slug": "GB",
        "something_else": "foo"
      },
      {
        "slug": "NI",
        "something_else": "bar"
      }
    ]
  }
  ```

---

> I've moved API views to `[app_name]/api_views.py` files, and serializers to `[app_name]/serializers.py`. This is a trend started in #633 – I think it's reasonable, but views welcome. This approach at the moment doesn't support versioning, so I wonder if I need to make them all packages with a sub-package per version? Not sure how to deal with this longer term, but I do like apps having their API gubbins in them somewhere.

I reckon storing them in `[app_name]/api/next/views.py` and `[app_name]/api/next/serializers.py` probably makes the most sense. As soon as we cut a version we need to do that anyway, so might as well start now..

--

> People now include all `PersonIdentifiers`, in a list rather than a hash / dict. This enables us to move to `n` IDs per (person, type) in future, but might make use tricky before then?

I think the way you've done it looks fine for now. If you're happy to go with that for now and come back to it, I reckon that's reasonable. I have had enough opinions about this for now ;)

symroe commented:
Comments on Chapter 1:

Good points, all. I've pushed some commits that I think addresses most things.

Points:

* I've created a base `CandidacySerializer` with `CandidacyOnBallotSerializer` and `CandidacyOnPersonSerializer` inheriting from it. This allows us to share the bulk of the code, but still lets us turn fields on and off with `BASE_CANDIDACY_FIELDS`, `CANDIDACY_ON_BALLOT_FIELDS` and `CANDIDACY_ON_PERSON_FIELDS`. I think this deals with most of the points, and allows us to share the code a bit

* I've added results to `CandidacyOnBallotSerializer` and also `BallotSerializer`, where relevant (e.g.  turnout on ballot, votes case on candidacy)

* I've added URLs to more things

I think this is an improvement overall, so thanks for picking up on them 👍

symroe commented:
## Chapter 2.1: `versions`, `history`

I reckon a sub-route makes sense here, so I've added it to ballot too. On discovery: the DRF interface shows "extra actions" in the header of the page by default, so that's one method of discovery. The docs would be another. I've added links to the routes on the endpoints directly – do you reckon that's alright?

I reckon adding filters to `LoggedAction` is good too, but mainly for filtering by type of action, etc.

symroe commented:
## Chapter 2: Party registers

Yes, `null` is a valid value for a party register:

1. We create some parties that aren't "real". "Speaker seeking re-election" and "independent" for example.

2. There are parties in the EC register that don't have a country register. Strictly, I think these are all party types that can't stand candidates

Good points on the consistency – I think I've addresses all your points

symroe commented:
Ok, a load of stuff done on this. The major thing to do is to move the code in to `[app]/api/[version]/[filename]` and add some more filters in. 

I'll crack on with that tomorrow before you start looking at the actual code.

Then, I think we're about ready to get the tests working again

chris48s commented:
## Chapter 3

Not sure what book this is we're writing JSON and the Argonauts maybe? JSON and his Technicolour Dream API? :D


In general, this is heading in a great direction. Couple of minor easy points first:

---

* LoggedAction object should have a ballot FK (side issue: this should also be in the front-end)
* Same bit of feedback about election_types I gave on party registers: http://127.0.0.1:8000/api/next/election_types.json throws an error. Wrap it up in a paginated results object - same as everything else.

---

I think the biggest issue we've got now is the way we're presenting results.
Seeing the latest iteration this feels very wierd. At the moment, each candidate has a `result` object `{"is_winner", "num_ballots"}`, but then the ballot also has a `results` object which has  `{"num_turnout_reported", "num_spoilt_ballots", "source"}` (but doesn't contain actual results). At a first glance its non-obvious what `result` and `results` do and why they're different. `num_turnout_reported` and `num_spoilt_ballots` seem like they could just be fields on the ballot, but I see that it is useful to group them with `source`, but then the `is_winner` and `num_ballots` fields in the other `result` objects are also from that same `source`, but we've lost that connection. I can put the jigsaw back together, but its not very intuitive. Imagine trying to document this, or explain this design decision to someone who pops up in our slack who is interested in using the YNR API to get election results.
Happy to hear your thoughts on it, but having seen this I think it makes more sense to serialize a subset of result info into the ballot/person and just link to a single `/results/{ballot_id}` endpoint for the detail which groups all this into a nice object similar to the existing `result_sets` endpoint/object.


---

The final issue for me is the `NULL` party register. Once we add a `?reigster=` filter to the parties endpoint (which we haven't done yet), there's going to be some parties which aren't reachable using that filter because `?reigster=` will query where `register == ''` and `?reigster=null` will query where `register == 'null'`. There will be no way for the user to specify where `register is None`, so I think we need to assign that a value at some point and explain what its for. I'm happy to punt that one down the line for the moment though if you want.

---


This last pile of comments I've just been looking at the JSON outputs. I've not really got into the actual code at all yet. Chapter 4 (in which our protagonist looks at the diff) can come next week..

symroe commented:
> makes more sense to serialize a subset of result info into the ballot/person and just link to a single /results/{ballot_id} endpoint for the detail which groups all this into a nice object similar to the existing result_sets endpoint/object.

I'm partly convinced that a dedicated endpoint might be handy, as some people basically only want to see results, but with this approach we can filter on `has_results=True` (and other things), and we get the same data returned.

I don't really see the addition of the results on the data as a problem, so I'd be interested to know what you think it wrong with it – let's chat about this on a call.

It's possible that we settle on one short term solution and create a new issue for the API V1 milestone. This would allow us to get some feedback on how confusing it is.

symroe commented:
> The final issue for me is the NULL party register.

I've punted this to [an issue](#1055) for the API v1 milestone – as we can't (and don't need to) currently filter on `Party` it's not really a new problem in this PR.

chris48s commented:
## Chapter 4

More working title suggestions for our novel:

* "To kill a `unittest.Mock`-ing bird"
* "`try/catch`-er in the Rye"

Perhaps "The curious incident of the docs in the night time" is a good title for this point in the review ;)


Bunch of comments inline. Beyond that..

---

There's a mix of documentation pages at URLs like

* `/api/terms`
* `/api/csv`

and then some under `/api/docs/`

It feels like this it would be nicer to keep docs out of the base `/api` namespace and have them all under `/api/docs/` with just a redirect on the root of `/api` to `/api/docs`. If this is hard because `/docs` is the auto-generated stuff and `/api/terms` etc are the "hand-crafted" docs, what would you think about moving hand-crafted docs to `/api/docs` and putting the auto-generated bit in `/api/docs/spec` or `/api/docs/openapi` or something? I know this is a bit bikesheddy, but the URL schema is one of those things that is harder to change later once we've shipped it.

---

There are a bunch of endpoints which have the wrong description on them:


`/ballots/{ballot_paper_id}/history/` says 

> A paginated list of all ballots
> Returns a Ballot object

That's the description for `/ballots/`.


`/election_types/` and `/election_types/{id}/` both say

> Return a list of all party register values, used for discovery of filter values in the `parties` endpoint.

That's the description for `/party_registers/`


`/party_registers/{id}/` says

> Return a list of all party register values, used for discovery of filter values in the parties endpoint.

That's also the description for `/party_registers/`. This should be 

> Returns a `PartyRegister` object


`/people/{id}/history/` and `/people/{id}/versions/` both say

> Returns a Person object

That's the description for `/people/{id}/`

---

Really the core thing in the `/ballots` output as a user is the `candidacies` array, Knowing what the structure of a candidacy object is like is pretty key to getting value out of this, but if you look the the docs for `candidacies` under a Ballot object
http://127.0.0.1:8000/api/docs/next/definitions/#Ballot/candidacies
its just documented (incorrectly) as a string. Also the same for the equivalent on a Person object: http://127.0.0.1:8000/api/docs/next/definitions/#Person/candidacies
I reckon having accurate definitions for that is quite important.

---

There are some ballot filters like `review_required`, `has_sopn` which we're describing as "string" filters, but really they're booleans. I know they're a bit odd because they're GET params. Is there any simple way to classify these better? We can come back to this if its not really simple.

---

There are quite a lot of places where we're describing an int field as having a minimum of `-2147483648` where it only makes sense for them to be 0 or positive. e.g: turnout, spoiled ballots, identifiers, etc:

![Screenshot at 2019-10-29 11-26-24](https://user-images.githubusercontent.com/6025893/67769729-e7819b00-fa4c-11e9-8aeb-45af61fec814.png)

If there's an easy way for us to say these have a minimum of zero, that would be really useful. If not, we can punt it to an issue.

---

In general, all of the definitions (except for `Party` which has nice useful descriptions) need a bunch of work doing to add descriptions explaining semantics. We can go with this for now, but they do need adding at some point. Hopefully now that we've got some structure in place, adding those descriptions would make a good volunteer task if someone cares..

---

I kind of get where you're going with the diagram, but I reckon it needs some work:

* Its clear that a ballot has many candidacies, but the diagram implies "a ballot has many candidacies" is the only one-to-many, whereas in reality an organisation can have many elections, an election contains many ballots, an organisation has many divisions, etc.
* The diagram implies each candidacy has a `Result` object, but then if you look at the definition for a `Result` object http://127.0.0.1:8000/api/docs/next/definitions/#Result that's the `Result` that relates to a ballot.
* I'm not really sure what the different shaped box means for ballots.
* I can see why you've put Ballot in the middle, but my instinct is to read the diagram top-to-bottom and in that context, Organisation seems like an odd place to start. I feel like we could probably make this clearer by essentially flipping it on its side and repositioning ballot in the middle at the top (so its the obvious first node of the tree).

In general, a diagram is a useful addition to the docs here, but even as someone who is already familiar with the data model, I wasn't really sure what to make of it. What do you reckon to dropping this iteration of the diagram for initial deploy and putting together an improved version for a future iteration?

symroe commented on ynr/apps/api/templates/api/next_home.html:
```diff
@@ -0,0 +1,111 @@
+
+The project contains data for all general election candidates since 2010, scheduled election candidates for local, police and crime commissioners, devolved governments and mayoral elections since 2016. See the [Election Types][#election-types] section for more information about elections covered.
+
+## TODO
```
https://github.com/DemocracyClub/yournextrepresentative/issues/1067

symroe commented on ynr/apps/api/templates/api/csv_docs.html:
```diff
@@ -0,0 +1,26 @@
+  There is a row per candidacy, so if a single person has stood on 3 ballots, they will have 3 rows in the file.
+  </p>
+
+  <p><a href="{{ MEDIA_URL }}candidates-all.csv" class="button">Download all the candidates</a></p>
```
We link to the CSV on the `Election` (not `Ballot`) pages, e.g. https://candidates.democracyclub.org.uk/elections/local.warwick.2019-10-29/

The _real_ solution here is something along the lines of the interface I prototyped in #962, but that's for another day

chris48s replied:
This is one of those things where neither of us have any user research to back what we're saying here, but I feel like there is probably a class of "spreadsheet warrior" user who is going to discover the CSV downloads via the front-end, but if the discovery point is via a JSON endpoint, they become essentially invisible.

I agree that the optimal long-term solution here isn't "have a list of >1,000 individual CSVs", but given there's going to be a fairly substantial chunk of time between now and when we arrive at a better solution for that (during which time we're probably going to have more eyes on this than any other time), I think we should provide a stop-gap solution here for finding CSVs via the front-end by reproducing the list we currently generate on https://candidates.democracyclub.org.uk/help/api _somewhere_ .

chris48s replied:
Related: @JoeMitchell  has literally just blogged

> Looking for a spreadsheet of the candidates? That’ll be on [this page](https://candidates.democracyclub.org.uk/help/api)

at https://democracyclub.org.uk/blog/2019/10/29/its-on/

I think that is a good example of a situation where its useful to have a stopgap for a more human-facing way to expose these.

chris48s replied:
True. Although in the case of a local election, that would have to be..

https://candidates.democracyclub.org.uk/elections/local.adur.2016-05-05/
https://candidates.democracyclub.org.uk/elections/local.amber-valley.2016-05-05/
https://candidates.democracyclub.org.uk/elections/local.barnsley.2016-05-05/
...`<snip>`...
https://candidates.democracyclub.org.uk/elections/local.worcester.2016-05-05
https://candidates.democracyclub.org.uk/elections/local.worthing.2016-05-05
https://candidates.democracyclub.org.uk/elections/local.wyre-forest.2016-05-05

..although I accept 1 page with some-hundred links isn't that much nicer than some-hundred pages in that case.

Also remember there are chunks of elections where the page https://candidates.democracyclub.org.uk/elections/parl.2017-06-08/ (for example) _does exist_ , but its not actually linked anywhere because its in the past so its much harder to discover than https://candidates.democracyclub.org.uk/elections/parl.2019-12-12/

Either way, I'm 100% on board with replacing big list o'CSVs with something better but given we don't have something better yet, it seems replacing it with nothing in the meantime is a step back here.

If you're committed to removing it, at least mentioning the two workarounds/discovery mechanisms we've discussed here:

* CSVs are linked from the `/elections` JSON endpoint and
* CSVs are linked from the `/elections` HTML pages

in the documentation for CSVs would be a massive improvement over what is there now.

chris48s commented on ynr/apps/uk_results/api/next/api_views.py:
```diff
@@ -0,0 +1,28 @@
+
+class ResultViewSet(viewsets.ReadOnlyModelViewSet):
+    lookup_field = "ballot__ballot_paper_id"
+    lookup_value_regex = "[^/]+"
```
Should this be `lookup_value_regex = r"(?!\.json$)[^/]+"` ?

chris48s commented on ynr/apps/elections/api/next/api_views.py:
```diff
@@ -0,0 +1,164 @@
+    """
+
+    lookup_field = "ballot_paper_id"
+    lookup_value_regex = "[^/]+"
```
Should this be `lookup_value_regex = r"(?!\.json$)[^/]+"` ?

chris48s commented on ynr/apps/elections/api/next/api_views.py:
```diff
@@ -0,0 +1,164 @@
+from utils.db import LastWord
+
+
+class UpcomingElectionsView(View):
```
Is this view actually mounted on any route? There's a separate version of this view in the `v0.9` codebase to cover the `v0.9` endpoint. If its unreachable code, we can just bin it.

chris48s commented on ynr/apps/elections/api/next/api_views.py:
```diff
@@ -0,0 +1,164 @@
+                    "-elected",
+                    "-result__is_winner",
+                    "-result__num_ballots",
+                    "person__sort_name",
```
Would it make sense to sort by

```
...
"-result__num_ballots",
"party_list_position",
"person__sort_name",
...
```

here so that `party_list_position` takes precedence over name if we have no results?

symroe commented:
# Chapter 4: Integers

I've changed all of these (I think) to a `PositiveIntegerField` or `PositiveSmallIntegerField`. This addresses the issue, and has the benefit of tightening up our data model

symroe commented on ynr/apps/elections/filters.py:
```diff
@@ -76,6 +114,49 @@ class Meta:
+    has_results = HasResultsFilter(label="Has Results", widget=AnyBooleanWidget)
+
+
+class CurrentOrFutureBallotFilter(BaseBallotFilter):
```
That's correct – do you want me to add a doc string or something?

chris48s replied:
No I was just asking to clear up my own understanding.

chris48s commented on ynr/apps/api/templates/api/terms.html:
```diff
@@ -0,0 +1,31 @@
+{% filter markdown %}
+This means you are free to reuse the data for any reason, as long as you give attribution to "Democracy Club"
+and any reuse is shared under the same licence.
+[Please read the full terms for more information](https://creativecommons.org/licenses/by-sa/3.0/).
```
Lets update this to

```markdown
This means you are free to reuse the data for any reason, as long as you give attribution to "Democracy Club".
[Please read the full terms for more information](https://creativecommons.org/licenses/by/4.0/).
```

chris48s commented on ynr/apps/api/templates/api/api-home.html:
```diff
@@ -0,0 +1,40 @@
+
+### Usage terms
+
+* The data on this site is provided under the [Attribution-ShareAlike 3.0 (CC BY-SA 3.0) licence](https://creativecommons.org/licenses/by-sa/3.0/).
```
```suggestion
* The data on this site is provided under the [Attribution 4.0 (CC BY 4.0) licence](https://creativecommons.org/licenses/by/4.0/).
```

chris48s replied:
```suggestion
* The data on this site is provided under the [Attribution 4.0 (CC BY 4.0) licence](https://creativecommons.org/licenses/by/4.0/).
```

chris48s commented on ynr/apps/api/templates/api/terms.html:
```diff
@@ -0,0 +1,31 @@
+{% block api_content %}
+<h1>Data reuse terms</h1>
+{% filter markdown %}
+This means you are free to reuse the data for any reason, as long as you give attribution to "Democracy Club"
```
```suggestion
This means you are free to reuse the data for any reason, as long as you give attribution to "Democracy Club".
```

chris48s commented on ynr/apps/api/templates/api/terms.html:
```diff
@@ -0,0 +1,31 @@
+<h1>Data reuse terms</h1>
+{% filter markdown %}
+This means you are free to reuse the data for any reason, as long as you give attribution to "Democracy Club"
+and any reuse is shared under the same licence.
```
```suggestion
```

chris48s commented on ynr/apps/api/templates/api/terms.html:
```diff
@@ -0,0 +1,31 @@
+  or from sources where it seems reasonable that we might use them on this site,
+  such as the party's official page for that candidate, or their social media profile picture.
+
+If the CC-BY-SA licence is problematic for you, but you feel that you have a worthwhile use of
```
```suggestion
If the CC-BY-A licence is problematic for you, but you feel that you have a worthwhile use of
```

symroe commented:
> We're now describing it as an object instead of a string

Yes. This took me ages of faffing about last night before I decided to abandon ship. It's one of those "falling off the edge of the framework" problems, where it's very hard to change. I'd like to punt this to an issue TBH.

chris48s commented:
> I'd like to punt this to an issue TBH

Yeah - understood. FWIW, documenting nested/inline objects and arrays of them was one of the things I spent a long time getting right with the devs.DC documentation too. In general, I think this is one of the higher priority docs issues, but it doesn't have to block the rest of this for now.

One of the nice things about defining docs using something like API Blueprint/Open API/Core API is that you can generate a schema from the docs. At the moment, the outputs from our API wouldn't validate against the schema generated from our own docs. I reckon getting to a stage where that is true is a fairly important target.

symroe commented:
I've pushed a commit that documents the embedded objects properly now – the table is getting quite ugly, but it's less wrong now. 

I still can't display the `object` as an `array`, and I think it will need a PR to `drf-yasg`, as the code producing this has no hooks to overwrite, and it burred deep in the logic.

The quick overview:

When using a `@action` decorator on a ViewSet you can use `detail=True|False`. If it's `True` then DRF passes a PK, assuming it's a route for a single object. If it's `False` you don't get a PK. Because We're returning an array _based on a PK_, we need to use `detail=True`, but DRF-YASG looks at this flag and sets the return value to an `object`.

One solution to this (that might be useful in general) would be to return an actual object with a list of elements in – this would allow us to e.g. page the results at some point, or add more fields.

chris48s commented:
OK. Its strange that `results` can be a "List of `Ballot` objects", but `candidacies` can't be a "List of `InlineCandidacy` objects", but I'm not deep enough into YASG to have any useful input here. I'll defer to your expertise on it :+1: Either way, its an issue for another day.

btw, one more thing (sorry!) - I just had another look at this and realised that all this is done for `Ballot.candidacies` but not `Person.candidacies` which still says its a string

chris48s commented:
...and they all lived happily ever after ;)
